### PR TITLE
Use bcryptjs instead of brcypt to avoid issues with compiling

### PR DESCRIPTION
**Pull request by alexandertrefz**
bcryptjs is API compatible to bcrypt but 100% pure JS, which avoids issues with the node-gyp compiling, for example having the wrong python version and similar issues.

Basically, it is the same thing but just JS, so no issues caused by the native stuff, which makes for better forward and backwards compatibility and an easier time when deploying. Always bet on JS™ :)

### CODE REVIEW

**Comment by nickjj** [comment]
Thanks for the pull request. Have you used bcryptjs for a while? It looks like a fairly new library with not much of a following. Cryptography is really important and I would be weary of just switching over to it without hearing some success stories of it being used out in the wild.

**Comment by alexandertrefz** [comment]
TL;DR: No i have not used it, but think the module is fine.

Well, it has compatibility to bcrypt and the tests worked nicely with the exception that "123" was not accepted as a hash and it threw an error on that.
I have not used that library yet, but i am also no cryptography expert, so even if i had, i could not tell you if it was good or not. If you know your crypto you can see if the crypto is sound or not. If you don't, and most developers including me, and i guess you as well, have to trust that it works. [This](http://pilif.github.io/2013/07/why-I-dont-touch-crypto/) sums up my thoughts on this pretty exactly.

What i can say is that the guy who wrote this has done quite a couple of modules that are quite low level JavaScript and some have a somewhat of a following. So i just put my trust in him, just like you put the trust in the author of the bcrypt module, that is does its job correct(since the 45000+ people that downloaded that modules surely are not crypto experts). But since the algorithm is something that was often implement, and he only reimplemented the algorithm and did not change something about it, i think the module is fine.

**Comment by julianlam** [comment]
Fun note -- if you search "bcrypt" in github, this issue comes up, so that's how I found this.

My two cents -- both [NodeBB](//github.com/designcreateplay/NodeBB) and [Ghost](https://github.com/TryGhost/Ghost) are using `bcryptjs` in production, with only one side-effect: Raspberry Pis choke on the actual hashing step, but if that's not a common use case for your project, this is a non-issue.

Ghost specifically recommends you replace "bcryptjs" with "bcrypt" for Raspberry Pis, so the workaround is fairly straightforward. Hashes are 1:1, can just drop-in replace.

**Comment by nickjj** [comment]
I originally threw together this module for a project being hosted on a low end VPS (ec2 micro instance).

I seriously doubt in any reasonable production environment for a web app your bottleneck will be due to using the JS version of bcrypt however my concern is what **real** benefits do you gain by using the js version?

If you provision an app server once every 3 months due to growth and compiling bcrypt takes 30 seconds but your entire app server takes 40 minutes to provision from start to finish then we're only talking about a 1.25% decrease in server provision time that you gain once every 3 months by going with the js version.

The downside is each time you call the lib it's less efficient. This may happen 50 or 100 times a day and it affects the response time of your end users. I'm fully aware this is a non-issue 99.999% of the time and you'll likely be IO bound, etc..

If you guys can throw out some real benefits of switching to the js version I'll happily merge it.

**Comment by alexandertrefz** [comment]
- You do not need to compile anything.

That is well enough of a reason in my opinion - not because you save some fraction of time, but because your requirements go down. You can use the module in more places. As written in the first post, i for one would like to use a particular hoster that fits my needs well, but i do not have control over the python version installed, and thus the bcrypt module wont compile. bcryptjs just works. No hassle. That is a huge improvement in my eyes.

The time that the actual bcrypt-calls take is irrelevant, actually, you need to keep it high, with your hardware growing, that is the whole point -> to stay safe against dictionary-attacks you want your basic encryption-time to be high. In effect that means you want your encryption to take at least 1-2s because having a user wait 1s is not a high cost for actual safety, but having to wait 1s for each tested password makes dictionary attacks take years.
That makes fighting over efficiency and saved time quite pointless. It is all about having more compatibility.

**Comment by nickjj** [comment]
May I ask what host you're using and what version of python it has which is causing the problem?

It never crossed my mind that someone couldn't even compile the bcrypt module so I didn't really relate not having to compile anything as a benefit because if your host could compile it, the compile time ends up being close to nothing in terms of time spent provisioning a server.

**Comment by julianlam** [comment]
Again, it really depends on use case, but for our projects, after switching to bcryptjs, we no longer had to require users to install `build-essential`, since nothing was being compiled anymore.

Of course, in your case, that may be moot as `build-essential` may be used for other things :smile:
